File: ci/data_extractor/src/formatters/integer/integer.py
import collections
import enum

from benchmark_specs import (
    ALL_RUST_INTEGER_TYPES,
    Backend,
    BenchDetails,
    BenchType,
    OperandType,
    RustType,
    ZKOperation,
)
from formatters.common import (
    OPERATION_SIZE_COLUMN_HEADER,
    BenchArray,
    GenericFormatter,
    ZKGenericFormatter,
)


class OperationDisplayName(enum.StrEnum):
    Negation = "Negation (-)"
    AddSub = "Add / Sub (+,-)"
    Mul = "Mul (x)"
    EqualNotEqual = "Equal / Not Equal (eq, ne)"
    Comparisons = "Comparisons (ge, gt, le, lt)"
    MaxMin = "Max / Min (max, min)"
    Bitwise = "Bitwise operations (&, \\|, ^)"
    Div = "Div  (/)"
    Rem = "Rem  (%)"
    DivRem = "Div / Rem  (/, %)"
    Shifts = "Left / Right Shifts (<<, >>)"
    Rotates = "Left / Right Rotations (left_rotate, right_rotate)"
    LeadingTrailing = "Leading / Trailing zeros/ones"
    Log2 = "Log2"
    Select = "Select"


class IntegerFormatter(GenericFormatter):
    @staticmethod
    def _format_data(data: dict[BenchDetails : list[int]], conversion_func):
        formatted = collections.defaultdict(
            lambda: {
                2: "N/A",
                8: "N/A",
                16: "N/A",
                32: "N/A",
                64: "N/A",
                128: "N/A",
                256: "N/A",
            }
        )
        for details, timings in data.items():
            test_name = "_".join((details.sign_flavor.value, details.operation_name))
            bit_width = details.bit_size
            value = conversion_func(timings[-1])

            if bit_width == 40:
                # Ignore this width as it's not displayed publicly.
                continue

            formatted[test_name][bit_width] = value

        return formatted

    def _generate_arrays(
        self,
        data,
        operand_type: OperandType = None,
        included_types: list[RustType] = ALL_RUST_INTEGER_TYPES,
        excluded_types: list[RustType] = None,
    ):
        match operand_type:
            case OperandType.CipherText:
                prefix = "unsigned"
            case OperandType.PlainText:
                prefix = "unsigned_scalar"

        match self.backend:
            case Backend.CPU:
                operations = [
                    f"{prefix}_neg_parallelized",
                    f"{prefix}_add_parallelized",
                    f"{prefix}_mul_parallelized",
                    f"{prefix}_eq_parallelized",
                    f"{prefix}_gt_parallelized",
                    f"{prefix}_max_parallelized",
                    f"{prefix}_bitand_parallelized",
                    f"{prefix}_div_rem_parallelized",
                    f"{prefix}_left_shift_parallelized",
                    f"{prefix}_rotate_left_parallelized",
                    f"{prefix}_leading_zeros_parallelized",
                    f"{prefix}_ilog2_parallelized",
                    f"{prefix}_if_then_else_parallelized",
                ]
            case Backend.GPU:
                operations = [
                    f"{prefix}_neg",
                    f"{prefix}_add",
                    f"{prefix}_mul",
                    f"{prefix}_eq",
                    f"{prefix}_gt",
                    f"{prefix}_max",
                    f"{prefix}_bitand",
                    f"{prefix}_div_rem",
                    f"{prefix}_left_shift",
                    f"{prefix}_rotate_left",
                    f"{prefix}_leading_zeros",
                    f"{prefix}_ilog2",
                    f"{prefix}_if_then_else",
                ]
            case Backend.HPU:
                operations = [
                    f"{prefix}_sub",  # Negation operation doesn't exist in HPU yet
                    (
                        f"{prefix}_add"
                        if operand_type == OperandType.CipherText
                        else f"{prefix}_adds"
                    ),
                    (
                        f"{prefix}_mul"
                        if operand_type == OperandType.CipherText
                        else f"{prefix}_muls"
                    ),
                    f"{prefix}_cmp_eq",
                    f"{prefix}_cmp_gt",
                    f"{prefix}_max",
                    f"{prefix}_bw_and",
                    (
                        f"{prefix}_div"
                        if operand_type == OperandType.CipherText
                        else f"{prefix}_divs"
                    ),
                    (
                        f"{prefix}_shift_l"
                        if operand_type == OperandType.CipherText
                        else f"{prefix}_shifts_l"
                    ),
                    (
                        f"{prefix}_rot_l"
                        if operand_type == OperandType.CipherText
                        else f"{prefix}_rots_l"
                    ),
                    f"{prefix}_lead0",
                    f"{prefix}_ilog2",
                    f"{prefix}_if_then_else",
                ]
            case _:
                raise NotImplementedError(
                    f"backend '{self.backend}' not supported yet for integer formatting"
                )

        display_names = [
            OperationDisplayName.Negation,
            OperationDisplayName.AddSub,
            OperationDisplayName.Mul,
            OperationDisplayName.EqualNotEqual,
            OperationDisplayName.Comparisons,
            OperationDisplayName.MaxMin,
            OperationDisplayName.Bitwise,
            OperationDisplayName.DivRem,
            OperationDisplayName.Shifts,
            OperationDisplayName.Rotates,
            OperationDisplayName.LeadingTrailing,
            OperationDisplayName.Log2,
            OperationDisplayName.Select,
        ]

        types = included_types.copy()
        excluded_types = excluded_types if excluded_types is not None else []
        for excluded in excluded_types:
            types.remove(excluded)

        first_column_header = OPERATION_SIZE_COLUMN_HEADER

        # Adapt list to plaintext benchmarks results.
        if operand_type == OperandType.PlainText and self.backend != Backend.HPU:
            if self.backend == Backend.CPU:
                div_name = f"{prefix}_div_parallelized"
                rem_name = f"{prefix}_rem_parallelized"
            elif self.backend == Backend.GPU:
                div_name = f"{prefix}_div"
                rem_name = f"{prefix}_rem"

            operations.insert(8, div_name)
            operations.insert(9, rem_name)
            operations.pop(7)

            display_names.insert(
                8,
                OperationDisplayName.Div,
            )
            display_names.insert(
                9,
                OperationDisplayName.Rem,
            )
            display_names.pop(7)  # Remove Div / Rem

            # Negation operation doesn't exist in plaintext
            operations.pop(0)
            display_names.pop(0)

        data_without_excluded_types = {}
        for op, values in data.items():
            try:
                data_without_excluded_types[op] = {
                    typ: val
                    for typ, val in values.items()
                    if RustType.from_int(typ) in types
                }
            except NotImplementedError:
                # Unknown type from database, ignoring
                continue

        filtered_data = filter(lambda t: t in operations, data_without_excluded_types)
        # Get operation names as key of the dict to ease fetching
        filtered_data_dict = {
            item: tuple(data_without_excluded_types[item].values())
            for item in filtered_data
        }

        result_lines = []
        for name, op in zip(display_names, operations):
            try:
                line = {first_column_header: name.value}
                line.update(
                    {
                        types[i].name: value
                        for i, value in enumerate(filtered_data_dict[op])
                    }
                )
                result_lines.append(line)
            except KeyError:
                # Operation not found in the results, ignoring this line.
                print(
                    f"backend '{self.backend}' could not find operation '{op}' to put in line '{name}'"
                )
                continue

        return [
            BenchArray(result_lines, self.layer),
        ]


class ZKFormatter(ZKGenericFormatter):
    @staticmethod
    def _get_default_dict() -> collections.defaultdict:
        return collections.defaultdict(
            lambda: {
                ZKOperation.Proof: "N/A",
                ZKOperation.Verify: "N/A",
                ZKOperation.VerifyAndExpand: "N/A",
            }
        )

    @staticmethod
    def _match_case_variation_filter(*args, **kwargs):
        # At this layer, server-like ZK are performed there are no variations such as browser kind.
        # Simply match all cases.
        return True

File: utils/tfhe-data-extractor/src/format/integer.rs
//! The `operation × ciphertext size` table of the integer layer.

use benchmark_spec::{
    Backend, BenchPath, IntegerBench, IntegerOp, IntegerOpBySign, OperandType, TfheLayer,
};

use super::{Cells, GridSpec, Measured, Table, build_grid, readable_value};

const ROW_HEADER: &str = r"Operation \ Size";

/// Published precisions. FheUint2, FheUint4 and FheUint256 are benchmarked but
/// not shown.
const COLUMNS: &[(&str, i64)] = &[
    ("FheUint8", 8),
    ("FheUint16", 16),
    ("FheUint32", 32),
    ("FheUint64", 64),
    ("FheUint128", 128),
];

// Spelled once so the six catalogues below cannot drift apart.
const NEGATION: &str = "Negation (-)";
const ADD_SUB: &str = "Add / Sub (+,-)";
const MUL: &str = "Mul (x)";
const EQUALITY: &str = "Equal / Not Equal (eq, ne)";
const COMPARISONS: &str = "Comparisons (ge, gt, le, lt)";
const MAX_MIN: &str = "Max / Min (max, min)";
const BITWISE: &str = "Bitwise operations (&, |, ^)";
const DIV_REM: &str = "Div / Rem  (/, %)";
const DIV: &str = "Div  (/)";
const REM: &str = "Rem  (%)";
const SHIFTS: &str = "Left / Right Shifts (<<, >>)";
const ROTATIONS: &str = "Left / Right Rotations (left_rotate, right_rotate)";
const LEADING_TRAILING: &str = "Leading / Trailing zeros/ones";
const LOG2: &str = "Log2";
const SELECT: &str = "Select";

/// Rows in publication order. A label may name several operations while only
/// one is benchmarked: `Add / Sub` is measured on the addition alone.
const CPU_CIPHERTEXT_ROWS: &[(&str, IntegerOp)] = &[
    (NEGATION, IntegerOp::NegParallelized),
    (ADD_SUB, IntegerOp::AddParallelized),
    (MUL, IntegerOp::MulParallelized),
    (EQUALITY, IntegerOp::EqParallelized),
    (COMPARISONS, IntegerOp::GtParallelized),
    (MAX_MIN, IntegerOp::MaxParallelized),
    (BITWISE, IntegerOp::BitandParallelized),
    (DIV_REM, IntegerOp::DivRemParallelized),
    (SHIFTS, IntegerOp::LeftShiftParallelized),
    (ROTATIONS, IntegerOp::RotateLeftParallelized),
    (LEADING_TRAILING, IntegerOp::LeadingZerosParallelized),
    (LOG2, IntegerOp::Ilog2Parallelized),
    (SELECT, IntegerOp::IfThenElseParallelized),
];

/// Same, for `scalar` operations, with `Div / Rem` split in two.
///
/// Negation, leading and trailing zeros/ones, and log2 are missing because they
/// are unary: there is no scalar form to measure. `Select` is missing for a
/// different reason, being available on scalars but not benchmarked, so its row
/// belongs here again the day it is.
const CPU_SCALAR_ROWS: &[(&str, IntegerOp)] = &[
    (ADD_SUB, IntegerOp::ScalarAddParallelized),
    (MUL, IntegerOp::ScalarMulParallelized),
    (EQUALITY, IntegerOp::ScalarEqParallelized),
    (COMPARISONS, IntegerOp::ScalarGtParallelized),
    (MAX_MIN, IntegerOp::ScalarMaxParallelized),
    (BITWISE, IntegerOp::ScalarBitandParallelized),
    (DIV, IntegerOp::ScalarDivParallelized),
    (REM, IntegerOp::ScalarRemParallelized),
    (SHIFTS, IntegerOp::ScalarLeftShiftParallelized),
    (ROTATIONS, IntegerOp::ScalarRotateLeftParallelized),
];

/// The Cuda benches call the unparallelized names: the parallelism is the
/// device's, not rayon's.
const GPU_CIPHERTEXT_ROWS: &[(&str, IntegerOp)] = &[
    (NEGATION, IntegerOp::Neg),
    (ADD_SUB, IntegerOp::Add),
    (MUL, IntegerOp::Mul),
    (EQUALITY, IntegerOp::Eq),
    (COMPARISONS, IntegerOp::Gt),
    (MAX_MIN, IntegerOp::Max),
    (BITWISE, IntegerOp::Bitand),
    (DIV_REM, IntegerOp::DivRem),
    (SHIFTS, IntegerOp::LeftShift),
    (ROTATIONS, IntegerOp::RotateLeft),
    (LEADING_TRAILING, IntegerOp::LeadingZeros),
    (LOG2, IntegerOp::Ilog2),
    (SELECT, IntegerOp::IfThenElse),
];

const GPU_SCALAR_ROWS: &[(&str, IntegerOp)] = &[
    (ADD_SUB, IntegerOp::ScalarAdd),
    (MUL, IntegerOp::ScalarMul),
    (EQUALITY, IntegerOp::ScalarEq),
    (COMPARISONS, IntegerOp::ScalarGt),
    (MAX_MIN, IntegerOp::ScalarMax),
    (BITWISE, IntegerOp::ScalarBitand),
    (DIV, IntegerOp::ScalarDiv),
    (REM, IntegerOp::ScalarRem),
    (SHIFTS, IntegerOp::ScalarLeftShift),
    (ROTATIONS, IntegerOp::ScalarRotateLeft),
];

/// Two rows are looser than their label: negation does not exist on the device
/// yet and is measured on the subtraction, and only the division is measured
/// behind `Div / Rem`. Both are how the published tables already read.
const HPU_CIPHERTEXT_ROWS: &[(&str, IntegerOp)] = &[
    (NEGATION, IntegerOp::Sub),
    (ADD_SUB, IntegerOp::Add),
    (MUL, IntegerOp::Mul),
    (EQUALITY, IntegerOp::CmpEq),
    (COMPARISONS, IntegerOp::CmpGt),
    (MAX_MIN, IntegerOp::Max),
    (BITWISE, IntegerOp::BwAnd),
    (DIV_REM, IntegerOp::Div),
    (SHIFTS, IntegerOp::ShiftL),
    (ROTATIONS, IntegerOp::RotL),
    (LEADING_TRAILING, IntegerOp::Lead0),
    (LOG2, IntegerOp::Ilog2),
    (SELECT, IntegerOp::IfThenElse),
];

/// Shorter than the others: the spec has no scalar spelling for equality,
/// comparison, bitwise, leading zeros, log2 or select on this backend.
const HPU_SCALAR_ROWS: &[(&str, IntegerOp)] = &[
    (NEGATION, IntegerOp::ScalarSub),
    (ADD_SUB, IntegerOp::ScalarAdds),
    (MUL, IntegerOp::ScalarMuls),
    (MAX_MIN, IntegerOp::ScalarMax),
    (DIV_REM, IntegerOp::ScalarDivs),
    (SHIFTS, IntegerOp::ScalarShiftsL),
    (ROTATIONS, IntegerOp::ScalarRotsL),
];

fn rows(backend: Backend, operand: OperandType) -> &'static [(&'static str, IntegerOp)] {
    match (backend, operand) {
        (Backend::Cpu, OperandType::CipherText) => CPU_CIPHERTEXT_ROWS,
        (Backend::Cpu, OperandType::PlainText) => CPU_SCALAR_ROWS,
        (Backend::Cuda, OperandType::CipherText) => GPU_CIPHERTEXT_ROWS,
        (Backend::Cuda, OperandType::PlainText) => GPU_SCALAR_ROWS,
        (Backend::Hpu, OperandType::CipherText) => HPU_CIPHERTEXT_ROWS,
        (Backend::Hpu, OperandType::PlainText) => HPU_SCALAR_ROWS,
    }
}

/// Builds the integer table for one backend and operand type.
///
/// The backend comes from the caller: a parsed id does not carry it, and every
/// row of a run shares the one the query pinned.
pub fn table(measured: &[Measured], backend: Backend, operand: OperandType) -> Table {
    let table_rows = rows(backend, operand);
    let mut cells: Cells<(IntegerOp, i64)> = Cells::new();

    for m in measured {
        // Unsigned operations only. The catalogue alone tells the ciphertext
        // and scalar tables apart: scalar benches are their own operations
        // (`scalar_add_parallelized`), and `operand_type` is never set on an
        // integer bench, so it cannot discriminate.
        let BenchPath::Tfhe(TfheLayer::Integer(IntegerBench::Ops(IntegerOpBySign::Unsigned(op)))) =
            m.spec.bench_crate()
        else {
            continue;
        };
        cells.insert((op, m.bit_size), readable_value(m.spec.metric(), m.value));
    }

    let (grid, empty_rows) = build_grid(
        GridSpec {
            row_header: ROW_HEADER,
            rows: table_rows
                .iter()
                .map(|(label, op)| (label.to_string(), *op))
                .collect(),
            columns: COLUMNS
                .iter()
                .map(|(name, key)| (name.to_string(), *key))
                .collect(),
            // An operation with no result at all is left out: the row list is a
            // catalogue, and benchmarking a subset of it is routine.
            drop_empty_rows: true,
        },
        |op, size| cells.get(&(*op, *size)).cloned(),
    );

    Table {
        grid,
        empty_rows,
        conflicts: cells
            .conflicts()
            .map(|c| {
                let (op, bits) = c.key;
                format!("{op}::{bits} bits ({} / {})", c.kept, c.dropped)
            })
            .collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_label_never_names_two_operations_within_one_catalogue() {
        for backend in [Backend::Cpu, Backend::Cuda, Backend::Hpu] {
            for operand in [OperandType::CipherText, OperandType::PlainText] {
                let catalogue = rows(backend, operand);
                for (index, (label, _)) in catalogue.iter().enumerate() {
                    assert!(
                        !catalogue[..index].iter().any(|(seen, _)| seen == label),
                        "{backend:?}/{operand:?} lists {label:?} twice",
                    );
                }
            }
        }
    }

    /// The scalar catalogues must hold scalar operations, or the `-plaintext`
    /// table silently republishes the ciphertext figures.
    #[test]
    fn scalar_catalogues_only_hold_scalar_operations() {
        for backend in [Backend::Cpu, Backend::Cuda, Backend::Hpu] {
            for (label, op) in rows(backend, OperandType::PlainText) {
                assert!(
                    op.to_string().starts_with("scalar_"),
                    "{backend:?} measures {label:?} from {op}, which is not a scalar operation",
                );
            }
        }
    }
}
